docs/design/single-file-unsupported.md

```diff
@@ -0,0 +1,68 @@
+# Single-file unsupported attribute
+
+Starting with .NET Core 3.0 users have the ability to bundle all files used by their applications into a single executable through the publish functionality, `PublishSingleFile`. Although most applications can be agnostic of whether they will be deployed as single-file, there exist some APIs that are known to be incompatible with this deployment mode and thus can cause single-file apps to crash.
+
+While the existing single-file analyzer can help us emit warnings whenever incompatible runtime library APIs are called from within user's code in a project that has been marked with the intent of being published as single-file (through setting `PublishSingleFile` to true in its .csproj), developers can still take a dependency on third-party components implementing functionality that inadvertently make their application no longer compatible with single-file.
+
+## User Experience
+
+For library authors, the new attribute will allow them to easily annotate APIs known to be problematic when used in the context of a single-file application.
```

Comment:
This does not address the problem for library authors. How are the library authors going to know where to add the attribute?

Reply:
Agreed that we should expand on the experience for library authors - something like:
* Low level APIs will be annotated by us (CodeBase for example) - or will be directly recognized by analyzers
* Library authors will run analyzers on their code and potentially get warnings in places where the library code has issues (by getting warnings from the above)
* They can choose to either fix the library so that it works regardless of single-file
* Or if not possible or desirable, they can use the new attribute to annotate the affected functionality as problematic with single-file

The open question is how will library authors declare the intent to support single-file. This is similar to https://github.com/dotnet/sdk/issues/14562, but libraries are not published typically.
For the purposes of this discussion I would simply say that the library developer can build/publish with /p:PublishSingleFile=true to get the diagnsotics.

Reply:
The experience for library authors will need to be discussed some more - as using the properties has some issues for libraries (unlike apps, where it is much more fitting). Related to https://github.com/dotnet/runtime/issues/43540 and also https://github.com/dotnet/runtime/issues/43543 (which has a similar problem for trimming).

Reply:
> How are the library authors going to know where to add the attribute?

We already have https://docs.microsoft.com/en-us/dotnet/core/deploying/single-file listing the different APIs that are not compatible with single-file. All these are public APIs that will be recognized by the single-file analyzer, so a method that calls any of these would trigger a warning and the developer should either find a way to fix it or annotate the caller method as `SingleFileUnsupported`.

I agree with @vitek-karas that the tricky part is how will library developers state the intent of wanting their app to be compatible with single-file. I think that for now we can move on with what's proposed in the linked issue, so developers can put `PublishSingleFile=true` in their csproj and the single-file analyzer will warn them whenever an incompatible API is used.

Reply:
Can we include the list of APIs which will be annotated with this attribute in the proposal?
Is the plan to annotate also PNSE throwing APIs with this attribute (feels like an odd match) ?

Reply:
I'm not sure about the PNSE throwing APIs - I think it will depend on the specific API. If the main reason why they throw is the fact that the target platform/configuration is single-file, then I think it should be annotated with the attribute. If the main reason is something else, and it just so happens to also affect single-file-ness, maybe the existing platform support annotation should be used.

It also depends on where should PNSE be used. Do we really only use it for things which are just impossible to do on the specific platform (architecture or OS), or do we use it also for cases where the configuration of the app makes it impossible (single-file, trimming, AOT, ...).

Reply:
I guess PNSE should really be RID/platform-specific, otherwise, there is NotSupportedException

Reply:
> Can we include the list of APIs which will be annotated with this attribute in the proposal?

I've added them explicitly in the proposal and mentioned them in this doc as the first thing we know we want to annotate with this new attribute.